tests/test_browser.go

```diff
@@ -0,0 +1,297 @@
+/*
+ *
+ * xk6-browser - a browser automation extension for k6
+ * Copyright (C) 2021 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package tests
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"strconv"
+	"testing"
+
+	"github.com/dop251/goja"
+	"github.com/grafana/xk6-browser/api"
+	"github.com/grafana/xk6-browser/chromium"
+	"github.com/oxtoacart/bpool"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	k6common "go.k6.io/k6/js/common"
+	k6http "go.k6.io/k6/js/modules/k6/http"
+	k6lib "go.k6.io/k6/lib"
+	k6metrics "go.k6.io/k6/lib/metrics"
+	k6test "go.k6.io/k6/lib/testutils/httpmultibin"
+	k6stats "go.k6.io/k6/stats"
+	"gopkg.in/guregu/null.v3"
+)
+
+// browser is a test browser for integration testing.
+type browser struct {
+	ctx     context.Context
+	rt      *goja.Runtime
+	state   *k6lib.State
+	http    *k6test.HTTPMultiBin
+	samples chan k6stats.SampleContainer
+	api.Browser
+}
+
+// testBrowser configures and launches a chrome browser.
+// It automatically closes the browser when `t` returns.
+//
+// opts provides a way to customize the testBrowser.
+// see: withLaunchOptions for an example.
+func testBrowser(t testing.TB, opts ...interface{}) *browser {
```

Comment:
There are no instances in Go, only values 😅 Jokes aside, it's not like "test the browser," but like having a "test browser." Mostly, there won't be a stutter in practice because these are unexported so nobody will call them like `tests.testBrowser` or `tests.newTestBrowser`. 

`testBrowser` func brings us a new test browser which we use to test a browser: so it's natural to say `testBrowser` which returns a `browser`, only for testing. I like to keep it concise but I can change it to: `newTestBrowser() *testBrowser` . No worries.

> Though really, this would be clearer if all of these helpers were in the tests/helpers package, instead of directly in tests, as they're not tests. So this could be called with helpers.newTestBrowser().

I think what you suggest is unnecessarily verbose and there is no need to have a second package because we're using them only for testing. Each test in the test pkg needs the browser test type. It's not a helper, it's the way all our integration tests work now!

As I said before, using names like helper, util, etc. is an anti-pattern and bad style of naming a package. If you don't believe me, please take a look at this, from https://go.dev/blog/package-names:

> Avoid meaningless package names. Packages named util, common, or misc provide clients with no sense of what the package contains. This makes it harder for clients to use the package and makes it harder for maintainers to keep the package focused. Over time, they accumulate dependencies that can make compilation significantly and unnecessarily slower, especially in large programs. And since such package names are generic, they are more likely to collide with other packages imported by client code, forcing clients to invent names to distinguish them.
> 
> The name of the package is a critical piece of its design. Work to eliminate meaningless package names from your projects.

Also [this](https://about.sourcegraph.com/go/idiomatic-go/#Package-util), [this](https://m.youtube.com/watch?v=ltqV6pDKZD8&t=734s), and [this](https://twitter.com/inancgumus/status/1468300005060579328).

Avoid meaningless package names like util, common, misc, api, types, and interfaces from here:
https://github.com/golang/go/wiki/CodeReviewComments#package-names

From [Rob Pike](https://talks.golang.org/2012/splash.article):
* Verbosity can inhibit readability by obscuring the intent.
* Be concise while remaining expressive.
* Package names are usually concise.

Go is expressive, concise, clean, and efficient ([this](https://go.dev/doc/)).

Good package names should be concise and named after their purpose. Good pkgs are coherent in what they do and provide. Utils, utility, common, helper, or API are all non-idiomatic and non-effective pkg names. It's like saying I don't know where to put these.

It's better to put things where they are used instead of inventing unnecessary pkgs like utils or helpers etc. Some benefits are writing concise code (_which is one of the tenets of writing effective code in Go_). Pkg names and their usage should be concise and even abbreviated if it doesn't delude its purpose.

Reply:
Let's not retread the same package naming discussion. You don't have to appeal to authority to get your point across. Like I said in the other comment, conventions should be considered pragmatically, rather than as a strict rule to always follow. Otherwise they would be part of `gofmt` or a linter and we wouldn't be having this discussion.

My main issue with `testBrowser()` is that it doesn't describe what the function does, which has been a constant "convention" in all programming languages. While we're linking things, you might want to read [this](https://www.cs.utah.edu/~germain/PPS/Topics/functions.html) (couldn't find a better reference :man_shrugging:):

> The Name - describes the purpose of the function. Usually a verb or phrase, such as "compute_Average", or just "average".

It's why all test functions in Go begin with `Test*`, why most constructors begin with `New*`, etc. Because it most accurately describes the purpose of the function. `testBrowser()` implies that it's testing `browser`, when it's actually constructing a `browser` instance, so `newBrowser()` is more intuitive.

It feels kind of silly discussing these basic details, so if you feel strongly about `testBrowser()` then let's keep it as is. As far as suggestion levels go, this one is not high up on my list. :)

The reason I prefer a separate "helpers" package is that when I want to look at tests only, it's clear that this should be in `*_test.go`. I definitely don't want to see any functions or structs there that are not tests themselves. So in this sense `browser` is a helper struct, even if all our tests depend on it, and not a test itself. I would also want it to be reachable by tests only, even if it's currently private, as the current situation of having `tests/browser_test.go` and `tests/test_browser.go` is confusing.

Anyway, I don't have strong suggestions about improving this, as I didn't see the previous structure as a big problem we needed to address. So whatever your decision is, I'm fine with it. :)

Reply:
I think I already agreed and changed it to `newTestBrowser`, previously.

> I definitely don't want to see any functions or structs there that are not tests themselves.

I pragmatically follow conventions in the stdlib and community. So it's OK for me to have so-called _helper_ funcs and structs besides tests. A test should have whatever code is necessary for testing code under test closer to it, not in another package—important for coherence as well.

> I would also want it to be reachable by tests only, even if it's currently private.

It is now—only reachable by tests because every file in that directory is a test.

> having tests/browser_test.go and tests/test_browser.go is confusing.

You can think of all the things that don't end with a `_test.go` file a helper. Win-win. You made me hate the word helper and utility 😂

> Anyway, I don't have strong suggestions about improving this, as I didn't see the previous structure as a big problem we needed to address. So whatever your decision is, I'm fine with it. :)

Thanks!